src/lib/support/tests/TestUtf8.cpp

```diff
@@ -0,0 +1,170 @@
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 0)));
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 1)));
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 2)));
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 3)));
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 4)));
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(zero, 16)));
+    }
+
+    {
+        char insideZero[] = "test\0zero";
+        NL_TEST_ASSERT(inSuite, Utf8::IsValid(CharSpan(insideZero)));
+    }
+}
+
+#define TEST_INVALID_BYTES(...)                                                                                                    \
+    {                                                                                                                              \
+        uint8_t _buff[] = { __VA_ARGS__ };                                                                                         \
+        CharSpan _span(reinterpret_cast<const char *>(_buff), sizeof(_buff));                                                      \
+        NL_TEST_ASSERT(inSuite, !Utf8::IsValid(_span));                                                                            \
+    }                                                                                                                              \
+    (void) 0
+
+void TestInvalidStrings(nlTestSuite * inSuite, void * inContext)
+{
+    // overly long representation
+    TEST_INVALID_BYTES(0xe0, 0b1001'1111, 0x80); // A
+    TEST_INVALID_BYTES(0xed, 0b1011'0000, 0x80); // B
+    TEST_INVALID_BYTES(0xf0, 0b1000'1111, 0x80); // C
+
+    // Outside codepoint
+    TEST_INVALID_BYTES(0xf4, 0x90, 0x80, 0x80); // D
+    TEST_INVALID_BYTES(0xf4, 0x91, 0x82, 0x83);
+    TEST_INVALID_BYTES(0xf5, 0x81, 0x82, 0x83);
+
+    // Missing continuation
+    TEST_INVALID_BYTES(0xC2);
+    TEST_INVALID_BYTES(0xE0);
+    TEST_INVALID_BYTES(0xE1);
+    TEST_INVALID_BYTES(0xE1, 0x9F);
+    TEST_INVALID_BYTES(0xED, 0x9F);
+    TEST_INVALID_BYTES(0xEE, 0x9F);
+    TEST_INVALID_BYTES(0xF0);
+    TEST_INVALID_BYTES(0xF0, 0x9F);
+    TEST_INVALID_BYTES(0xF0, 0x9F, 0x9F);
+    TEST_INVALID_BYTES(0xF1);
+    TEST_INVALID_BYTES(0xF1, 0x9F);
+    TEST_INVALID_BYTES(0xF1, 0x9F, 0x9F);
+    TEST_INVALID_BYTES(0xF4);
+    TEST_INVALID_BYTES(0xF4, 0x9F);
+    TEST_INVALID_BYTES(0xF4, 0x9F, 0x9F);
+
+    // More tests from  https://www.w3.org/2001/06/utf-8-wrong/UTF-8-test.html
+    TEST_INVALID_BYTES(0x80); // First continuation byte
+    TEST_INVALID_BYTES(0xBF); // Last continuation byte
+
+    // Impossible bytes
+    TEST_INVALID_BYTES(0xFE);
+    TEST_INVALID_BYTES(0xFF);
+    TEST_INVALID_BYTES(0xFE, 0xFE, 0xFF, 0xFF);
+
+    // Overlong sequences
+    TEST_INVALID_BYTES(0xc0, 0xaf);
+    TEST_INVALID_BYTES(0xe0, 0x80, 0xaf);
+    TEST_INVALID_BYTES(0xf0, 0x80, 0x80, 0xaf);
+    TEST_INVALID_BYTES(0xf8, 0x80, 0x80, 0x80, 0xaf);
+    TEST_INVALID_BYTES(0xfc, 0x80, 0x80, 0x80, 0x80, 0xaf);
```

Comment:
These are invalid in different ways than overlong. They are technically invalid range UTF-16 codepoint range followed by unexpected continuations. Different implementations would consider these as different errors.

Suggest moving this to a different area, since they are still wrong, or commenting they are ambiguous.

Reply:
These were specifically picked up from w3 tests and I added the comments that they had.
I can remove the comment alltogether, but unsure if that will actually help. I would rather preserve these as-is ordered to match the source material I used as that is easier to reference.

Reply:
I added comments to hopefully understanding where the tests come from (and sligtly better grouping).

Also found a set of tests I missed, added those.